contrib/qubes/src/fwupd_receive_updates.py

```diff
@@ -0,0 +1,346 @@
+            print(
+                f'Domain {updatevm} not allowed to send dom0 updates',
+                file=sys.stderr
+            )
+            exit(1)
+
+    def _verify_received(self, files_path, regex_pattern, updatevm):
+        """Checks if sent files match  regex filename pattern.
+
+        Keyword arguments:
+
+        files_path -- absolute path to inspected directory
+        regex_pattern -- pattern of the expected files
+        updatevm - domain to be checked
+        """
+        for untrusted_f in os.listdir(files_path):
+            if not regex_pattern.match(untrusted_f):
+                raise Exception(f'Domain {updatevm} sent unexpected file')
+            f = untrusted_f
+            assert '/' not in f
+            assert '\0' not in f
+            assert '\x1b' not in f
+            path_f = path.join(files_path, f)
+            if os.path.islink(path_f) or not os.path.isfile(path_f):
+                raise Exception(f'Domain {updatevm} sent not regular file')
+
+    def _create_dirs(self, *args):
+        """Method creates directories.
+
+        Keyword arguments:
+        *args -- paths to be created
+        """
+        qubes_gid = grp.getgrnam('qubes').gr_gid
+        self.old_umask = os.umask(0o002)
+        if args is None:
+            raise Exception("Creating directories failed, no paths given.")
+        for file_path in args:
+            if not path.exists(file_path):
+                os.mkdir(file_path)
+                os.chown(file_path, -1, qubes_gid)
+                os.chmod(file_path, 0o0775)
+            elif os.stat(file_path).st_gid != qubes_gid:
+                print(
+                    f"{WARNING_COLOR}Warning: You should move a personal files"
+                    f" from {file_path}. Cleaning cache will cause lose of "
+                    f"the personal data!!{WARNING_COLOR}"
+                )
+
+    def _extract_archive(self, archive_path, output_path):
+        """Extracts archive file to the specified directory.
+
+        Keyword arguments:
+        archive_path -- absolute path to archive file
+        output_path -- absolute path to the output directory
+        """
+        cmd_extract = [
+            "cabextract",
+            "-d",
+            f"{output_path}",
+            f"{archive_path}"
+        ]
+        shutil.copy(archive_path, FWUPD_DOM0_UPDATES_DIR)
+        p = subprocess.Popen(cmd_extract, stdout=subprocess.PIPE)
+        p.communicate()[0].decode('ascii')
+        if p.returncode != 0:
+            raise Exception(
+                f'cabextract: Error while extracting {archive_path}.'
+            )
+
+    def _gpg_verification(self, file_path):
+        """Verifies GPG signature.
+
+        Keyword argument:
+        file_path -- absolute path to inspected file
+        """
+        cmd_gpg = [
+            "gpg",
+            "--verify",
+            f"{file_path}.asc",
+            f"{file_path}",
+        ]
```

Comment:
I think hardcoding GPG is probably dangerous; I think going forward we'd probably drop the standalone keys and just use the jcat file.

Reply:
It could be changed to `jcat-tool` if we will drop the Qubes R4.0 support. I'm not sure if there is `libjcat` package for Fedora 25.

Reply:
I've added the [`_jcat_verification`](https://github.com/fwupd/fwupd/pull/2710/files#diff-75406b905f82fd50a6d9b1cfe76b42b735a197d178d2cf9c4f1fc95fb2c16981R185) method, which is used for metadata verification.